### ISSUE TITLE
[Flow] add eslint-plugin-flowtype - require-valid-file-annotation

### ISSUE THREAD

**Issue by gregtatum** [devtools-html/debugger.html#5120]
It would be great to use this [eslint rule](https://github.com/gajus/eslint-plugin-flowtype#require-valid-file-annotation) to ensure that we don't mistakenly leave out `// @flow` headers.

Here are this list of files that are missing the header. Lets add the header and the minimal types needed to satisfy flow. It is okay to add a lot of loose types like `any`, `Object`, etc... 

### Example

here is an example patch that could be a PR

```diff
diff --git a/src/utils/ui.js b/src/utils/ui.js
index efc8b03..599989d 100644
--- a/src/utils/ui.js
+++ b/src/utils/ui.js
@@ -2,6 +2,8 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

+// @flow
+
 /* Checks to see if the root element is available and
  * if the element is visible. We check the width of the element
  * because it is more reliable than either checking a focus state or
@@ -14,7 +16,7 @@ export function isVisible() {

 /* Gets the line numbers width in the code editor
  */
-export function getLineNumberWidth(editor) {
+export function getLineNumberWidth(editor: Object) {
   const gutters = editor.display.gutters;
   const lineNumbers = gutters.querySelector(".CodeMirror-linenumbers");
   return lineNumbers && lineNumbers.clientWidth;
```

### Todo List

- [ ] src/actions
  - [ ] ast/setInScopeLines.js
  - [ ] breakpoints/addBreakpoint.js
  - [ ] breakpoints/remapLocations.js
  - [ ] event-listeners.js
  - [ ] source-tree.js
  - [ ] utils/middleware/log.js
  - [ ] utils/middleware/timing.js
- [ ] src/components
  - [ ] SecondaryPanes/Breakpoints/BreakpointsContextMenu.js
  - [ ] SecondaryPanes/Frames/types.js
- [ ] src/reducers
  - [ ] /async-requests.js
  - [ ] src/reducers/event-listeners.js
  - [ ] src/reducers/index.js
  - [ ] src/reducers/quick-open.js
  - [x] src/reducers/types.js
- [ ] src/selectors/
  - [ ] breakpointAtLocation.js
  - [ ] getCallStackFrames.js
  - [ ] isSelectedFrameVisible.js
  - [x] visibleBreakpoints.js
- [ ] src/utils
  - [x] DevToolsUtils.js
  - [x] bootstrap.js
  - [x] clipboard.js
  - [x] dbg.js
  - [ ] editor/source-editor.js (started here #5897)
  - [x] function.js
  - [x] indentation.js
  - [ ] location.js
  - [ ] project-search.js
  - [ ] result-list.js
  - [ ] source-maps.js
  - [ ] source-queue.js
  - [x] sources-tree/index.js
  - [ ] sources-tree/sortTree.js
  - [x] task.js
  - [ ] timings.js
  - [x] ui.js
  - [ ] worker.js
- [ ] src/global-types.js

**Comment by gregtatum** [comment]
Also, even if this was intentional, the eslint rule makes it explicit when you don't want to use flow in a file.

**Comment by wldcordeiro** [comment]
I noticed that several files are missing the header and I think they may have been that way because of some naive prop-types -> Flow conversions.

**Comment by jasonLaster** [comment]
yeah - we should definitely fix them!

**Comment by CameronPilgrim** [comment]
/claim
I am in @emmet0r 's class and my team ( @johnsonwe and @mohamedabo )would like to work on this issue.

**Comment by jasonLaster** [comment]
thanks! we might need to disable this for some files while we fix them, but not a big deal. when we do we can file issues to fix them!

**Comment by mohamedabo** [comment]
@jasonLaster  @gregtatum  Hello. My team has a few questions regarding this issue. What exactly do we need to do to implement the flow header?? Do we just add // @flow as said in the document? Also we have noticed that many files are missing this header. How do we identify which files need this header? (edited)

**Comment by jasonLaster** [comment]
yes, lets add them and then get `yarn flow` passing

**Comment by jasonLaster** [comment]
every JS file in `src` 
excluding spec.js and files in `src/test/mochitest`

**Comment by mohamedabo** [comment]
Hello @jasonLaster , my team have added the flow header to every JS file in src except spec.js files in src/test/mochitest. However, when I ran the command `yarn run flow` it gave me a list of errors showed in the following screen shots  
![screenshot from 2018-04-03 13-48-27](https://user-images.githubusercontent.com/32283283/38266259-2f8e3d20-3746-11e8-9854-4b8d3fbccf99.png) 
![screenshot from 2018-04-03 13-53-08](https://user-images.githubusercontent.com/32283283/38266317-5db93b8c-3746-11e8-8763-a9619c137ba2.png)

So basically my question is that what is our next step?

**Comment by jasonLaster** [comment]
@mohamedabo can you share the files that you added `// @flow` to. We have a bug for `reducers/types.js`

**Comment by mohamedabo** [comment]
Hi @jasonLaster. We already posted the commits. This is a list of the files that we have edited in our local repository https://github.com/johnsonwe/debugger.html/commit/a8c5abe199cdc5aaf6b5321233899d36b65196c8. If you needed an additional list let me know and I would be happy to provide it to you.

**Comment by jasonLaster** [comment]
Great @mohamedabo I just copied the files you touched into a todo list which is now in the description. Feel free to fix one file at a time and add a PR!

**Comment by atwalg2** [comment]
/claim I can claim the src/utils directory to start

**Comment by CameronPilgrim** [comment]
@mohamedabo, @johnsonwe, and I will tackle the src/components directory.

**Comment by mohamedabo** [comment]
@jasonLaster : @darkwing Told us that BreakpointsDropdown.js has been deleted. So, could you take that off the list? Thank you.

**Comment by mohamedabo** [comment]
Since the semester is getting to close, we are going to unclaim the issue. What we basically did was push a PR that was accepted and another PR that its file was deleted. Anyway, as stated above, the best course of action for anyone who wants to work on this project in the future is to add //@flow headers on top of these files and add flow types to the variables. /unclaim

**Comment by twhite96** [comment]
/claim
I'll get to work on this tonight. I am an Outreachy candidate. This seems like a good first issue.

**Comment by jasonLaster** [comment]
Oh great @twhite96. Welcome!

**Comment by akashs09** [comment]
/claim 

src/actions

**Comment by Kanikashrivastav** [comment]
@AnshulMalik i think this issue should not be closed , the latest PR added `Fixes #5120` due to that this issue is closed after merging that PR

**Comment by AnshulMalik** [comment]
Ohh, my bad, the keywords :P

**Comment by jasonLaster** [comment]
Here is a new list

we should likely add `"flowtype/require-valid-file-annotation": [2, "always"]` to our `eslintrc` and then disable these specific files so we can incrementally add support and not add new files that are missing flow

```
actions/ast/setInScopeLines.js
actions/breakpoints/breakpointPositions.js
actions/tests/helpers/breakpoints.js
actions/tests/helpers/readFixture.js
actions/tests/helpers/threadClient.js
actions/utils/middleware/log.js
actions/utils/middleware/timing.js
components/Editor/ColumnBreakpoints.js
components/Editor/EditorMenu.js
components/Editor/EmptyLines.js
components/Editor/menus/source.js
components/ProjectSearch.js
components/SecondaryPanes/Frames/FrameIndent.js
components/SecondaryPanes/Frames/Group.js
components/SecondaryPanes/Frames/types.js
components/shared/Button/index.js
components/shared/Popover.js
reducers/async-requests.js
reducers/index.js
reducers/quick-open.js
selectors/breakpointAtLocation.js
selectors/getCallStackFrames.js
selectors/isSelectedFrameVisible.js
selectors/visibleBreakpoints.js
selectors/visiblePausePoints.js
test/__mocks__/request-animation-frame.js
test/__mocks__/styleMock.js
test/__mocks__/svgMock.js
test/shim.js
test/tests-setup.js
test/utils/history.js
utils/editor/source-editor.js
utils/editor/token-events.js
utils/sources-tree/sortTree.js
utils/sources-tree/tests/addToTree.spec.js
utils/task.js
utils/url.js
utils/worker.js
vendors.js
workers/parser/tests/helpers/index.js
workers/parser/utils/formatSymbols.js
workers/search/get-matches.js
workers/search/index.js
workers/search/project-search.js
workers/search/worker.js
```

**Comment by nchevobbe** [comment]
Great, I'll do the ones in `Frames`

Le jeu. 31 janv. 2019 à 21:16, Jason Laster <notifications@github.com> a
écrit :
